kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoTelemetryListener.java

```diff
@@ -0,0 +1,139 @@
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.db.mongo;
+
+import com.mongodb.event.CommandFailedEvent;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandSucceededEvent;
+import dev.responsive.kafka.internal.metrics.ResponsiveMetrics;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.CumulativeCount;
+import org.apache.kafka.common.metrics.stats.CumulativeSum;
+
+/**
+ * Listener which reports MongoDB command failures and successes. For each result,
+ * we record a total count representing the number of commands executed which had that
+ * result (success or failure), and a cumulative latency. Reported metrics are tagged by
+ * the MongoDB command name.
+ */
+public class MongoTelemetryListener implements CommandListener {
+  private static final String MONGODB_METRICS_GROUP = "responsive-mongodb";
+
+  static final MetricName COMMANDS_SUCCEEDED_LATENCY = new MetricName(
+      "commands-succeeded-cumulative-latency",
+      MONGODB_METRICS_GROUP,
+      "cumulative commands succeeded latency",
+      Collections.emptyMap()
+  );
+  private static final MetricName COMMANDS_SUCCEEDED_COUNT = new MetricName(
+      "commands-succeeded-count",
+      MONGODB_METRICS_GROUP,
+      "total commands succeeded",
+      Collections.emptyMap()
+  );
+
+  static final MetricName COMMANDS_FAILED_COUNT = new MetricName(
+        "commands-failed-count",
+        MONGODB_METRICS_GROUP,
+        "total commands failed",
+        Collections.emptyMap()
+    );
+  static final MetricName COMMANDS_FAILED_LATENCY = new MetricName(
+      "commands-failed-cumulative-latency",
+      MONGODB_METRICS_GROUP,
+      "cumulative commands failed latency",
+      Collections.emptyMap()
+  );
+
+  private final ResponsiveMetrics metrics;
+
+  public MongoTelemetryListener(ResponsiveMetrics metrics) {
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void commandSucceeded(final CommandSucceededEvent event) {
+    getOrCreateSensor(event.getCommandName(), true)
+        .record(event.getElapsedTime(TimeUnit.MILLISECONDS));
+  }
+
+  @Override
+  public void commandFailed(final CommandFailedEvent event) {
+    getOrCreateSensor(event.getCommandName(), false)
+        .record(event.getElapsedTime(TimeUnit.MILLISECONDS));
+  }
+
+  private Sensor getOrCreateSensor(
+      String command,
+      boolean isSuccess
+  ) {
+    String sensorName = "mongodb-commands" + "-" + command
+        + (isSuccess ? "-succeeded" : "-failed");
+
+    Sensor sensor = metrics.getSensor(sensorName);
+    if (sensor == null) {
+      sensor = metrics.addSensor(sensorName);
+
+      if (isSuccess) {
+        sensor.add(commandsSucceededCount(command), new CumulativeCount());
+        sensor.add(commandsSucceededLatency(command), new CumulativeSum());
```

Comment:
I really need to brush up on my stats, but can we compute any histogram using htis or just average? I feel like p90/95/99 is helpful as well

Reply:
There is a `Percentiles` class included with Kafka metrics, but it is almost never used. Most of our histograms in Kafka tend to use the old yammer metrics implementation. Max is a straightforward (and cheap) addition though if you think we could get away with just that.

Reply:
I don't imagine `max` is super helpful for most of these - we can punt for now and see how far we get with average, that's what matters for throughput anyway.